File: src/connectivity/wlan/lib/common/rust/src/ie/wsc/id.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use zerocopy::{AsBytes, FromBytes, Unaligned};

#[repr(C, packed)]
#[derive(Eq, PartialEq, Hash, AsBytes, FromBytes, Unaligned, Copy, Clone, Debug)]
pub struct Id(pub [u8; 2]);

/// WSC Technical Specification v2.0.7, Section 12, Table 28
impl Id {
    pub const AP_SETUP_LOCKED: Self = Self([0x10, 0x57]);
    pub const CONFIG_METHODS: Self = Self([0x10, 0x08]);
    pub const DEVICE_NAME: Self = Self([0x10, 0x11]);
    pub const DEVICE_PASSWORD_ID: Self = Self([0x10, 0x12]);
    pub const MANUFACTURER: Self = Self([0x10, 0x21]);
    pub const MODEL_NAME: Self = Self([0x10, 0x23]);
    pub const MODEL_NUMBER: Self = Self([0x10, 0x24]);
    pub const PRIMARY_DEVICE_TYPE: Self = Self([0x10, 0x54]);
    pub const RESPONSE_TYPE: Self = Self([0x10, 0x3B]);
    pub const RF_BANDS: Self = Self([0x10, 0x3C]);
    pub const SELECTED_REG: Self = Self([0x10, 0x41]);
    pub const SELECTED_REG_CONFIG_METHODS: Self = Self([0x10, 0x53]);
    pub const SERIAL_NUMBER: Self = Self([0x10, 0x42]);
    pub const UUID_E: Self = Self([0x10, 0x47]);
    pub const VENDOR_EXT: Self = Self([0x10, 0x49]);
    pub const VERSION: Self = Self([0x10, 0x4A]);
    pub const WPS_STATE: Self = Self([0x10, 0x44]);
}

File: src/ui/scenic/lib/gfx/id.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/ui/scenic/lib/gfx/id.h"

#include <sstream>

namespace scenic_impl {

GlobalId::operator bool() const {
  static const GlobalId kNoGlobalId;
  return *this != kNoGlobalId;
}

GlobalId::operator std::string() const {
  std::ostringstream out;
  out << *this;
  return out.str();
}

bool operator==(const GlobalId& lhs, const GlobalId& rhs) {
  return lhs.session_id == rhs.session_id && lhs.resource_id == rhs.resource_id;
}

bool operator!=(const GlobalId& lhs, const GlobalId& rhs) { return !(lhs == rhs); }

bool operator<(const GlobalId& lhs, const GlobalId& rhs) {
  return lhs.session_id < rhs.session_id ||
         (lhs.session_id == rhs.session_id && lhs.resource_id < rhs.resource_id);
}

std::ostream& operator<<(std::ostream& os, const GlobalId& value) {
  return os << value.session_id << "-" << value.resource_id;
}

}  // namespace scenic_impl
